### PR TITLE
New release system

### LINKED ISSUE #2881: Reduce release frequency
After today's developer meeting we agreed to move to less frequent releases. The plan is to move to 3 releases a year (Jan, May, Sep) keeping calendar versioning (i.e. 23.1, 23.5 and 23.9). The release schedule from now should be 23.4 (end Apr), 23.5 (end May) and then 23.9 (end Sep).

Tasks:
- [ ] Fix release workflow
- [x] Announce new release schedule

### PR DESCRIPTION
**Pull request by Saransh-cpp**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #2881 

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [ ] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [ ] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [ ] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: .github/workflows/publish_pypi.yml [hunk 1]
```diff
@@ -22,7 +22,7 @@ jobs:
       - uses: actions/checkout@v3
       - uses: actions/setup-python@v4
         with:
-          python-version: 3.8
+          python-version: 3.11
```
Comment (Saransh-cpp) [inline .github/workflows/publish_pypi.yml:25]:
```suggestion
          python-version: 3.8
```

File: .github/workflows/publish_pypi.yml [hunk 2]
```diff
@@ -77,7 +77,7 @@ jobs:
...
```
Comment (Saransh-cpp) [inline .github/workflows/publish_pypi.yml:80]:
```suggestion
          python-version: 3.8
```

File: .github/workflows/publish_pypi.yml [hunk 3]
```diff
@@ -130,15 +130,12 @@ jobs:
...
-    strategy:
-      matrix:
-        python-version: [3.8]
 
     steps:
       - uses: actions/checkout@v3
       - uses: actions/setup-python@v4
         with:
-          python-version: ${{ matrix.python-version }}
+          python-version: 3.11
```
Comment (Saransh-cpp) [inline .github/workflows/publish_pypi.yml:138]:
```suggestion
          python-version: 3.8
```

File: .github/workflows/update_version.yml [hunk 4]
```diff
@@ -2,31 +2,48 @@ name: Update version
...
+        default: ""
+  schedule:
+     # Run at 10 am UTC on 1st every 4th month from January through December
+     - cron: 0 10 1 1-12/4 *
 
 jobs:
   update-version:
     # This workflow is only of value to PyBaMM and would always be skipped in forks
-    if: github.repository_owner == 'pybamm-team'
+    # if: github.repository_owner == 'pybamm-team'
```
Comment (Saransh-cpp) [inline .github/workflows/update_version.yml:16]:
```suggestion
    if: github.repository_owner == 'pybamm-team'
```

File: .github/workflows/update_version.yml [hunk 5]
```diff
@@ -2,31 +2,48 @@ name: Update version
...
+      - uses: actions/checkout@v3
+        if: github.event_name == 'workflow_dispatch'
+        with:
+          ref: 'main'
+
+      - name: Set up Python 3.11
         uses: actions/setup-python@v4
         with:
-          python-version: ${{ matrix.python-version }}
+          python-version: 3.11
```
Comment (Saransh-cpp) [inline .github/workflows/update_version.yml:33]:
```suggestion
          python-version: 3.8
```

**Review by agriyakhetarpal — CHANGES_REQUESTED** [review]
Just realised that these file paths were not changed earlier, it's good that we caught them before the 23.9 release:

https://github.com/pybamm-team/PyBaMM/blob/develop/scripts/update_version.py#L43-L47

should point to `docs/_static/versions.json` instead of `docs/source/_static/versions.json`

**Comment by Saransh-cpp** [comment]
This should be ready to review. Copying `release_workflow.md` here as this PR's description -

---

# Release workflow

 ## rc0 releases (automated)

 1. The `update_version.yml` workflow will run on every 1st of January, May and September, creating 2 PRs -

    1. Incrementing the version to `YY.MMrc0` by running `scripts/update_version.py` in the following files -
       - `pybamm/version.py`
       - `docs/conf.py`
       - `CITATION.cff`
       - `vcpkg.json`
       - `docs/source/_static/versions.json`

    2. A PR from `develop` to `main`

    The version PR should be merged into `develop`, and then the develop-to-main PR should be merged into `main`.

 2. Once the tests pass, create a new GitHub _pre-release_ with the same tag (`YY.MMrc0`) from the `main` branch and a description copied from `CHANGELOG.md`.

 3. This release will automatically trigger `publish_pypi.yml` and create a _pre-release_ on PyPI.

 ## rcX releases (manual)

 If a new release candidate is required after the release of `rc0` -

 1. Fix a bug in `main` (no new features should be added to `main` once `rc0` is released) and `develop` individually.

 2. Run `update_version.yml` manually while using `append_to_tag` to specify the release candidate version number (`rc1`, `rc2`, ...).

 3. This will create a PR incrementing the version to `YY.MMrcX` by running `scripts/update_version.py` in the following files -

    - `pybamm/version.py`
    - `docs/conf.py`
    - `CITATION.cff`
    - `vcpkg.json`
    - `docs/source/_static/versions.json`

    The version PR should be merged into `main`, because merging it into `develop` would require merging `develop` into `main`, something we don't want (`develop` will have new features).

 4. Once the tests pass, create a new GitHub _pre-release_ with the same tag from the `main` branch and a description copied from `CHANGELOG.md`.

 5. This release will automatically trigger `publish_pypi.yml` and create a _pre-release_ on PyPI.

 6. Manually merge `main` back to `develop` if any conflicts arise.

 ## Actual release (manual)

 Once satisfied with the release candidates -

 1. Run `update_version.yml` manually, leaving the `append_to_tag` field blank ("") for an actual release.

 2. This will create a PR incrementing the version to `YY.MM` by running `scripts/update_version.py` in the following files -

    - `pybamm/version.py`
    - `docs/conf.py`
    - `CITATION.cff`
    - `vcpkg.json`
    - `docs/source/_static/versions.json`

    The version PR should be merged into `main`, because merging it into `develop` would require merging `develop` into `main`, something we don't want (`develop` will have new features).

 3. Once the tests pass, create a new GitHub _release_ with the same tag from the `main` branch and a description copied from `CHANGELOG.md`.

 4. This release will automatically trigger `publish_pypi.yml` and create a _release_ on PyPI.

 ## Other checks

 Some other essential things to check throughout the release process -

 - Update baseline of registries in `vcpkg-configuration.json` as the latest commit id from [pybamm-team/sundials-vcpkg-registry](https://github.com/pybamm-team/sundials-vcpkg-registry)
 - Update `CHANGELOG.md` with a summary of the release
 - Update jax and jaxlib to the latest version in `pybamm.util` and fix any bugs that arise
 - If building wheels on Windows gives a `vcpkg` related error - revert the baseline of default-registry to a stable commit in `vcpkg-configuration.json`
 - Make sure the URLs in `docs/source/_static/versions.json` are valid

**Review by brosaplanella — APPROVED** [review]
Looks good to me, but let's wait to merge until the others are happy with it as this is quite a major point. Two main questions/comments:
1. Is the CHANGELOG updated automatically or do we need to do it manually (I couldn't quite tell)?
2. With this approach we will not have a separate branch for a specific version. Not sure we need it, but that's a major change from what we discussed in #2881 so just wanted to highlight it.

**Comment by Saransh-cpp** [comment]
Oh, I see, I missed incorporating a few points. I'll look into that. Thanks for the review!

File: .github/release_workflow.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+4. This release will automatically trigger `publish_pypi.yml` and create a _release_ on PyPI.
+
+## Other checks
+
+Some other essential things to check throughout the release process -
+
+- Update baseline of registries in `vcpkg-configuration.json` as the latest commit id from [pybamm-team/sundials-vcpkg-registry](https://github.com/pybamm-team/sundials-vcpkg-registry)
+- Update `CHANGELOG.md` with a summary of the release
+- Update jax and jaxlib to the latest version in `pybamm.util` and fix any bugs that arise
+- If building wheels on Windows gives a `vcpkg` related error - revert the baseline of default-registry to a stable commit in `vcpkg-configuration.json`
```
Comment (martinjrobins) [inline .github/release_workflow.md:75]: this is only really relevent if the `vcpkg` error is coming from new commits to `sundials-vcpkg-registry` or `cadadi-vcpkg-registry`. I would combine this and the first point as:

- If updating our custom `vcpkg` registory entries [pybamm-team/sundials-vcpkg-registry](https://github.com/pybamm-team/sundials-vcpkg-registry) or [pybamm-team/casadi-vcpkg-registry](https://github.com/pybamm-team/casadi-vcpkg-registry) (used to build Windows wheels), make sure to update the baseline of the registories in `vcpkg-configuration.json` to the latest commit id.

**Review by martinjrobins — CHANGES_REQUESTED** [review]
Looks great :) I just had one suggestion.

File: .github/release_workflow.md (hunk 6 above)
Comment (Saransh-cpp) [reply .github/release_workflow.md]: Thanks for the review, @martinjrobins! Fixed it! :)

**Comment by Saransh-cpp** [comment]
New logic with the branch stuff below! Here are some links from my fork -

- The [`v23.8` branch](https://github.com/Saransh-cpp/PyBaMM/tree/v23.8) - this should be created automatically on scheduled runs
- [Commit](https://github.com/Saransh-cpp/PyBaMM/commit/5a8f92370b84dde6e095a5d4abc136d1762faf64) updating the version to `rc1` - this was created by manually running `update_version.yml`
- [Commit](https://github.com/Saransh-cpp/PyBaMM/commit/4c0a76b30ca5e4ddcf50caca0a5b742dc936287d) updating the version to `v23.8` and a [PR](https://github.com/Saransh-cpp/PyBaMM/pull/12) from `23.8` to `main` - this was created by manually running `update_version.yml`

---

## rc0 releases (automated)

1. The `update_version.yml` workflow will run on every 1st of January, May and September, updating incrementing the version to `YY.MMrc0` by running `scripts/update_version.py` in the following files -

   - `pybamm/version.py`
   - `docs/conf.py`
   - `CITATION.cff`
   - `vcpkg.json`
   - `docs/_static/versions.json`

      These changes will be automatically pushed to a new branch `YY.MM`.

2. Create a new GitHub _pre-release_ with the tag `YY.MMrc0` from the `YY.MM` branch and a description copied from `CHANGELOG.md`.

3. This release will automatically trigger `publish_pypi.yml` and create a _pre-release_ on PyPI.

## rcX releases (manual)

If a new release candidate is required after the release of `rc0` -

1. Fix a bug in `YY.MM` (no new features should be added to `YY.MM` once `rc0` is released) and `develop` individually.

2. Run `update_version.yml` manually while using `append_to_tag` to specify the release candidate version number (`rc1`, `rc2`, ...).

3. This will increment the version to `YY.MMrcX` by running `scripts/update_version.py` in the following files -

   - `pybamm/version.py`
   - `docs/conf.py`
   - `CITATION.cff`
   - `vcpkg.json`
   - `docs/_static/versions.json`

      These changes will be automatically pushed to the existing branch `YY.MM`.

4. Create a new GitHub _pre-release_ with the same tag (`YY.MMrcX`) from the `YY.MM` branch and a description copied from `CHANGELOG.md`.

5. This release will automatically trigger `publish_pypi.yml` and create a _pre-release_ on PyPI.

## Actual release (manual)

Once satisfied with the release candidates -

1. Run `update_version.yml` manually, leaving the `append_to_tag` field blank ("") for an actual release.

2. This will increment the version to `YY.MMrcX` by running `scripts/update_version.py` in the following files -

   - `pybamm/version.py`
   - `docs/conf.py`
   - `CITATION.cff`
   - `vcpkg.json`
   - `docs/_static/versions.json`

      These changes will be automatically pushed to the existing branch `YY.MM`.

3. Next, a PR from `YY.MM` to `main` will be generated that should be merged once all the tests pass.

4. Create a new GitHub _release_ with the same tag from the `main` branch and a description copied from `CHANGELOG.md`.

5. This release will automatically trigger `publish_pypi.yml` and create a _release_ on PyPI.

## Other checks

Some other essential things to check throughout the release process -

- If updating our custom vcpkg registory entries [pybamm-team/sundials-vcpkg-registry](https://github.com/pybamm-team/sundials-vcpkg-registry) or [pybamm-team/casadi-vcpkg-registry](https://github.com/pybamm-team/casadi-vcpkg-registry) (used to build Windows wheels), make sure to update the baseline of the registories in vcpkg-configuration.json to the latest commit id.
- Update `CHANGELOG.md` with a summary of the release
- Update jax and jaxlib to the latest version in `pybamm.util` and `setup.py`, fixing any bugs that arise
- Make sure the URLs in `docs/_static/versions.json` are valid

File: .github/workflows/update_version.yml [hunk 7]
```diff
@@ -2,51 +2,73 @@ name: Update version
...
+        default: ""
+  schedule:
+     # Run at 10 am UTC on 1st every 4th month from January through December
+     - cron: 0 10 1 1-12/4 *
 
 jobs:
   update-version:
     # This workflow is only of value to PyBaMM and would always be skipped in forks
-    if: github.repository_owner == 'pybamm-team'
+    # if: github.repository_owner == 'pybamm-team'
```
Comment (agriyakhetarpal) [inline .github/workflows/update_version.yml:16]: I think you commented this out when you were testing the release on your fork 😄

File: scripts/update_version.py [hunk 8]
```diff
@@ -42,7 +40,7 @@ def update_version():
 
     # docs/source/_static/versions.json for readthedocs build
     with open(
-        os.path.join(pybamm.root_dir(), "docs", "source", "_static", "versions.json"),
+        os.path.join(pybamm.root_dir(), "docs", "_static", "versions.json"),
```
Comment (agriyakhetarpal) [inline scripts/update_version.py:43]: In one of the previous meetings @tinosulzer and I had decided not to enable `rcX` documentation builds and keep them for just stable releases. Could we add a check here too, with a conditional `if "rc" in release_version`? Also, a suggestion: I think splitting all the file-handling operations per file into their own separate functions would be great for improving readability

**Review by agriyakhetarpal — APPROVED** [review]
Thanks, @Saransh-cpp, this looks good! I added some minor review comments that you can address.

**Comment by agriyakhetarpal** [comment]
There is also a draft PR for adding `towncrier` (https://github.com/pybamm-team/PyBaMM/pull/3278) that we would need to address before the next release, just a heads up

**Review by brosaplanella — APPROVED** [review]
Looks good to me, thanks!

File: .github/workflows/update_version.yml [hunk 9]
```diff
@@ -2,51 +2,73 @@ name: Update version
...
+        default: ""
+  schedule:
+     # Run at 10 am UTC on day-of-month 1 in January, May, and September.
+     - cron: "0 10 1 1,5,9 *"
 
 jobs:
   update-version:
     # This workflow is only of value to PyBaMM and would always be skipped in forks
-    if: github.repository_owner == 'pybamm-team'
+    # if: github.repository_owner == 'pybamm-team'
```
Comment (Saransh-cpp) [inline .github/workflows/update_version.yml:16]:
```suggestion
    if: github.repository_owner == 'pybamm-team'
```

File: .github/workflows/release_reminder.yml [hunk 10]
```diff
@@ -0,0 +1,21 @@
+name: Create a release reminder
+
+on:
+  schedule:
+     # Run at 10 am UTC on 1st and 28th every 4th month from January through December
+     - cron: 0 10 1-28/27 1-12/4 *
```
[the remark this answers is not in the document]
Comment (Saransh-cpp) [reply .github/workflows/release_reminder.yml]: Thanks! That looks much better!

File: scripts/update_version.py (hunk 8 above)
Comment (Saransh-cpp) [reply scripts/update_version.py]: Oh yes, I remember that, thanks!

**Comment by Saransh-cpp** [comment]
Thanks for the reviews, @agriyakhetarpal and @brosaplanella!

**Comment by Saransh-cpp** [comment]
A gentle bump for a review on this, @martinjrobins. We would be able to test this new system if the PR is merged before 1st September.

**Review by martinjrobins — APPROVED** [review]
looks great, happy to merge

**Comment by Saransh-cpp** [comment]
Thanks, @martinjrobins!

**Pull request merged**